### PR TITLE
refactor(rest): refactor the proxy support for the REST Connector

### PR DESCRIPTION
**Pull request by johnBgood**
## Description

This PR changes the previous behavior about how we configure a proxy (and non-proxy hosts).
- We can now set up a proxy for REST-like connectors **only** (REST, Github, etc). This is done using a specific set of **environment variables**: 

| Env. Var | Description | Default |
|--------|--------|--------|
| CONNECTOR_HTTP(S)_PROXY_SCHEME | The proxy **scheme** (`http`, or `https`). This allows the user to use https proxies which is not possible with other clients | `http` |
| CONNECTOR_HTTP(S)_PROXY_HOST | The **host** part of the proxy URL (do not include `http://`)  | |
| CONNECTOR_HTTP(S)_PROXY_PORT | The **port** part of the proxy URL | |
| CONNECTOR_HTTP(S)_PROXY_USER | If you need to authenticate, use this variable to set the **user** | |
| CONNECTOR_HTTP(S)_PROXY_PASSWORD | If you need to authenticate, use this variable to set the **password** | | 
| CONNECTOR_HTTP_NON_PROXY_HOSTS | Similar to the [Java](https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html) `http.nonProxyHosts` system property | 

- Note that these values are **not copied** to the System Properties anymore.
- If you want a global behavior, don't use these variables but use System Properties instead.

## Checklist

- [x] PR has a **milestone** or the `no milestone` label.

### CODE REVIEW

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ public void build(ClassicRequestBuilder builder, HttpCommonRequest request) {
 
   String fetchOAuthToken(OAuthAuthentication authentication) {
     HttpCommonRequest oAuthRequest = oAuthService.createOAuthRequestFrom(authentication);
-    HttpCommonResult response = CustomApacheHttpClient.getDefault().execute(oAuthRequest);
+    HttpCommonResult response = new CustomApacheHttpClient().execute(oAuthRequest);
```
Comment (ztefanie) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: Could we create two clients (one for using a proxy and one for non-proxy) and then select the correct one for a request instead of creating new ones for every request?

**Review by ztefanie — COMMENTED** [review]
Feel free to re-add these tests for matching non proxy hosts properly:
https://github.com/camunda/connectors/pull/3894/files#diff-0d190eecc67fcce2f29db41ca7e5fbe821ff94693ab1ab045393c42e148abb91R91-R126

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java (hunk 1 above)
Comment (johnBgood) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: Sadly not really, the timeouts are configured when creating the client and they are based on the request values :/ 
```java
  private RequestConfig getRequestTimeoutConfig(TimeoutConfiguration timeoutConfiguration) {
    return RequestConfig.custom()
        .setConnectionRequestTimeout(
            Timeout.ofSeconds(timeoutConfiguration.connectionTimeoutInSeconds()))
        .setResponseTimeout(Timeout.ofSeconds(timeoutConfiguration.readTimeoutInSeconds()))
        .build();
  }
```
If you find another way, I'll be more than happy though

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java [hunk 2]
```diff
@@ -20,103 +20,49 @@
...
-    connectionManager.setDefaultSocketConfig(SocketConfig.custom().setSoKeepAlive(true).build());
-
-    return connectionManager;
-  }
-
   /**
    * Converts the given {@link HttpCommonRequest} to an Apache {@link
    * org.apache.hc.core5.http.ClassicHttpRequest} and executes it.
    *
+   * <p>This method is thread-safe.
```
Comment (ztefanie) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java:38]: is this `<p>` tag ever closed?

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java [hunk 3]
```diff
@@ -20,103 +20,49 @@
...
-              .build()
-              .execute(
-                  apacheRequest,
-                  new HttpCommonResultResponseHandler(
-                      executionEnvironment, request.isStoreResponse()));
-      if (HttpStatusHelper.isError(result.status())) {
-        throw ConnectorExceptionMapper.from(result);
+      var host = apacheRequest.getAuthority().getHostName();
+      var scheme = apacheRequest.getScheme();
+      try (var client =
```
Comment (ztefanie) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java:51]: Nice :rocket:

**Review by ztefanie — COMMENTED** [review]
Thanks for this PR and fixing the proxy issues :green_heart: 

I reviewed it, the code looks good :+1: 

I tested this, but the nonProxyHost setting gets ignored for me
These are my settings `CONNECTOR_HTTP_PROXY_HOST=localhost;CONNECTOR_HTTP_PROXY_PORT=3128;CONNECTOR_HTTP_PROXY_NON_PROXY_HOSTS=www.gmx.de;CONNECTOR_HTTP_PROXY_USER=my-user;CONNECTOR_HTTP_PROXY_PASSWORD=demo` but requests to http://www.gmx.de still get proxied.

Please also do not forget to adjust the documentation, plus I think we need to mention that we introduced breaking changes, because now a setup with proxy settings with system variables will no longer work, right?

Please also do not forget to backport this to 8.7

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java (hunk 1 above)
Comment (sbuettner) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: @johnBgood Curious, can this be overriden on a request level like this:
```
final RequestConfig params = RequestConfig.custom().setConnectTimeout(3000).setSocketTimeout(3000).build();
httpGet.setConfig(params);
```
?

Comment (sbuettner) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: It seems like the proxy config can also be passed via the request config: https://hc.apache.org/httpcomponents-client-4.5.x/current/httpclient/apidocs/org/apache/http/client/config/RequestConfig.html

Comment (johnBgood) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: > @johnBgood Curious, can this be overriden on a request level like this:

Yes, but this is not how we create requests (we don't use HttpGet, HttpPost etc). This is because it induced other limitations or less generic code I don't remember exactly

Comment (sbuettner) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: Ok, should we maybe tackle it in a separate PR? As our Connectors are essentially request based we could potentially benefit from a more request focused configuration.

Comment (johnBgood) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: Yes we can have a look in another PR, I took a look and it's probably not that many changes

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java (hunk 2 above)
Comment (johnBgood) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java:38]: no need to, it's usually not closed in Javadoc

**Review by ztefanie — APPROVED** [review]
Tested again, works as expected :rocket:

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Test required in (you can check both):
  - [ ] SaaS
  - [ ] SM
- [ ] Create a new test for this feature in our [Pre Release Tests](https://modeler.ultrawombat.com/projects/017779a8-9459-4feb-8b47-eeef8b2717ef--connectors-pre-release-test) for @Szik to be able to test

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java (hunk 1 above)
Comment (sbuettner) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/builder/parts/ApacheRequestAuthenticationBuilder.java:66]: Created a follow up issue here: https://github.com/camunda/connectors/issues/4371

**Review by Szik — APPROVED** [review]

**Review by ztefanie — APPROVED** [review]
